compose: Fix message lost during server downtime.

akshatdalton opened the pull request:
This commit fixes the bug: web app loses the messages sent during the server downtime.

Fixes #15692.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Testing plan:** <!-- How have you tested? -->
Added new test case.

**GIFs or screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


https://user-images.githubusercontent.com/67277428/108834488-40872200-75f4-11eb-8596-1d3a9f719ce1.mp4

You can skip to `3:45` when the localhost restarts.

timabbott commented on static/js/transmit.js:
```diff
@@ -15,7 +22,8 @@ exports.send_message = function (request, on_success, error) {
         error(xhr, error_type) {
-            if (error_type !== "timeout" && reload_state.is_pending()) {
+            if (error_type !== "timeout" || reload_state.is_pending()) {
+                compose_state.message_content(saved_message_content);
```
fab64fd7b0e78d7464f91003fb7d49007a9478b7 appears to where this bit of `&&` logic is introduced.  I don't think I understand the reasoning behind why we want an || gate here, though -- maybe worth adding a comment above this?

I **think** the idea is that if we're in the process of doing a reload, we guess -- but what does `error_type !== "timeout"` mean?

amanagr commented:
After shutting down the server and sending a message, I immediately get this:
<img width="1274" alt="Screenshot 2021-02-26 at 7 57 32 AM" src="https://user-images.githubusercontent.com/25124304/109247932-d1374b00-780a-11eb-8665-28d4db49b12e.png">
<img width="1274" alt="Screenshot 2021-02-26 at 8 14 42 AM" src="https://user-images.githubusercontent.com/25124304/109247937-d4cad200-780a-11eb-956a-c48e293909ae.png">

But on upstream/master, I am able to send messages when server is down.

akshatdalton commented:
> After shutting down the server and sending a message, I immediately get this:
> <img alt="Screenshot 2021-02-26 at 7 57 32 AM" width="1274" src="https://user-images.githubusercontent.com/25124304/109247932-d1374b00-780a-11eb-8665-28d4db49b12e.png">
> <img alt="Screenshot 2021-02-26 at 8 14 42 AM" width="1274" src="https://user-images.githubusercontent.com/25124304/109247937-d4cad200-780a-11eb-956a-c48e293909ae.png">
> 
> But on upstream/master, I am able to send messages when server is down.

It is expected as if we look at `initiate` function in `reload.js`, it changes the window url and then initiates a reload.

timabbott commented:
@Signior-X would you be up for doing your own manual testing of this PR?  I feel like you just did a bunch of the exact testing process for this sort of issue :)

Signior-X commented:
This PR needs to be closed as this issue has been fixed in #17359

Signior-X commented:
![image](https://user-images.githubusercontent.com/56730716/112394335-8aae1100-8d22-11eb-9b01-5459bcfe13fe.png)

When the server is down and I send a message, it get's like this. So, the thing is you can't send a message in server downtime. Now, the thing is as soon as I restart the server, and manually reload the page. The message was in the compose box. 

So, what I feel is this. It is preventing the message to be send in server downtime and getting the message back on restart.

Signior-X commented:
Is this can be related to #17697?
More or less, I feel they point to the same issue i.e., message lost during server downtime/network error have the same problem. The main aim is to prevent message from getting lost. As @timabbott suggested, we can have a button which can convert the messages to draft. Or we can have some function in reload, which will automatically adds the failed messages into the drafts list.... There are many possibilities here.

Also, is there any CZO link where we can have more discussion over this? I have one for the above issue [here](https://chat.zulip.org/#narrow/stream/2-general/topic/Failed.20message.20to.20draft.20.2317697)

akshatdalton commented:
@Signior-X, thanks for reviewing this PR in such great detail!

I was aware that after my changes, one couldn't continuously send the messages (during server downtime), and they won't be echoed. I actually focused more on preserving the message content from being lost after the server is up again and the web page reloads.

But after doing a revisit, I feel that messages (sent during server downtime) should be sent to the server (and probably I have found that path).

I feel that the changes in my PR are correct (up till now), but they need additional changes (from other files), I'll update my PR soon addressing those issues.

Signior-X commented:
Sure, @akshatdalton
Waiting for an update to this PR soon!

Signior-X commented:
So, I just reviewed the PR.
First, let me tell you how I checked the PR.
I started the server, opened localhost:9991, and then open dev tools -> network, and there changed to no throttling to offline.
Now, as I send any message, the same message that is being locally echoed (meaning just shown offline to the user, the message sent has failed), is copied into the compose box.
Let me add the video.

For the testing of the server-side problem, I switched off the server and it working as you said. Storing the last message in the compose box. This part is working fine, but the above case isn't.
I get the problem, there isn't any proper condition with which we can find the difference between the two cases. So, some other issue is being created by the PR.

akshatdalton commented:
@Signior-X, can you explain what do you mean by:
> but the above case isn't.
?

akshatdalton commented:
> I get the problem, there isn't any proper condition with which we can find the difference between the two cases.

If you are referring to that, this should only send a message when the server is down and not when the user is offline/lost connection; then we first need to discuss that in CZO if we want to do that only when the server is down. Identifying the issue is from the server or client-side is easy; I just need to add a function in `channel.js`.

Signior-X commented:
> @Signior-X, can you explain what do you mean by:
> 
> > but the above case isn't.
> 
> ?

By the above case, I mean client-side network error/failure.

>  Identifying the issue is from the server or client-side is easy; I just need to add a function in channel.js

Oh, that's good. Actually, we should keep the client-side error management as it is. 

> If you are referring to that, this should only send a message when the server is down and not when the user is offline/lost connection;

By sending a message what do you mean, the one that shows the message has failed and it is showing above or something else? 
 
My point of concern is to keep the client-side message failures as it is they are right now in master. To only handle the server down problem.

akshatdalton commented:
> > If you are referring to that, this should only send a message when the server is down and not when the user is offline/lost connection;
> 
> By sending a message what do you mean, the one that shows the message has failed and it is showing above or something else?

The one that has failed.

> My point of concern is to keep the client-side message failures as it is they are right now in master. To only handle the server down problem.

Any specific reason to not handle the client-side connection problem?
Same I have asked in: https://chat.zulip.org/#narrow/stream/9-issues/topic/Web.20app.20loses.20messages.20sent.20during.20server.20downtime.2E.20.2315692/near/1155623

timabbott commented:
@akshatdalton can you rebase this?

akshatdalton commented:
@timabbott, I can rebase this, but I don't all the work is done here w.r.t. new idea we discussed.
I left some comments in: https://chat.zulip.org/#narrow/stream/9-issues/topic/Web.20app.20loses.20messages.20sent.20during.20server.20downtime.2E.20.2315692/near/1155623
about how I'm planning to do this. Please have a look.
I'll soon complete this work and push my new changes.

akshatdalton commented:
I've rebased this, which fixes the issue **but** this is not what we have discussed in: https://chat.zulip.org/#narrow/stream/9-issues/topic/Web.20app.20loses.20messages.20sent.20during.20server.20downtime.2E.20.2315692/near/1155623.
I'll soon push the changes addressing the issues raised in [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Web.20app.20loses.20messages.20sent.20during.20server.20downtime.2E.20.2315692/near/1155623).

akshatdalton committed:
compose: Add css for `Send` button when it is disabled.

akshatdalton committed:
compose: Fix messages lost during server downtime.

When server is unreachable, which is possible
primarily because of two reasons:

1. client is facing some connection problem
2. server is down for maintenance checks

then in such a case if user tries to send message then
as a result it is failed to be acknowledged by the
server. When the connection is up again then after
the page has reloaded the sent messages are lost
and are not preserved either.
This commit fixes such issue.

We store these unsent messages (when server is unreachable)
in localStorage. We allow maximum 4 unsent messages to store.

When the connection is up again and page has reloaded, we
ask the user to acknowledge these unsent messages, if any.
They can either opt to send these messages or can cancel
them. This choice can be made on per message.

Fixes: #15692.

akshatdalton committed:
puppeteer: Add test for `unsent_messages.js`.

Here we use the Chrome DevTools property to go
offline and online. In the test, we send messages
while offline and when we are online again we
check that those messages, which were not responded
by the server, are asked to acknowledged by the user
in the order they were created or not.

akshatdalton committed:
Refactor: Remove unnecessary URL-encoding logic & use actual data structure in `reload.js`.

This commit completes a TODO in `static/js/reload.js`.
To preserve the states like pointer, narrow, message,
etc., `reload.js` was using unnecessary URL-encoding
logic, which is now removed and this now just passes
the actual data structure to the localStorage.
Same changes are done in the `initialize` function to
reflect the same behaviour.
This makes the codebase cleaner and more readable.

This also fixes a typo in regex in `initialize`
function, because of which even if initially the
parameter: `send_after_reload` was `true` in
`preserve_state` function, message was not sent
after the reload is initialized.

akshatdalton committed:
compose_state: Extract helper function: `construct_message_data`.

This commit extracts helper function: `construct_message_data` to
be used by `drafts.js` and `reload.js`. This logic was earlier
used by `drafts.js` only.

This will also act as a prep commit for future changes in this
related area. Related issue: #15692.

akshatdalton commented on frontend_tests/puppeteer_tests/unsent-messages.ts:
```diff
@@ -0,0 +1,134 @@
+
+    // Going online to send the last unsent messages.
+    await set_throttling_property(page, false);
+    await page.reload();
```
The Page will not reload immediately after the connection is up again, so we need to do a quick manual reload.

akshatdalton commented:
I've added few prep commits while working in this direction.
This PR aims to fix this issue in the following way:
1. When the server is unreachable (can be possible because of client connection problem or when the server is down for maintenance) and sends some messages (which will be failed to be acknowledged by the server), we store them in the localStorage. 
Right now, we allow up to a maximum of 4 unsent messages to be stored. However, we can exceed this limit up to a maximum of 99 unsent messages -- since any message that is not acknowledged by the server has a local id = maximum id + 0.01. So to avoid messing up the order in which they are sent, we shouldn't cross this limit.

2. After the server is reachable again and the page has reloaded, we ask the user to acknowledge the last unsent messages. Then, they can either opt to send it or can cancel them (can do this per message). All this is done through a template that pops up above the compose box.

This helps to cover most of the necessary conditions but leave us with some edge cases -
1. If a user has closed the tab and reloads in a fresh tab then the 1st unsent message is saved as a draft. This is because the compose box gets closed as the user is narrowed to the `Recent topics` section (right after reload). In this case, we should not let the compose box close.

2. Another case is when a message containing wildcard mention is not sent (because the server was unreachable), the next time after the page has reloaded, it is sent if a user clicks on `Yes, send` (in `compose_unsent_message` template).

Thanks to @Riken-Shah for providing initial support in the Puppeteer test!
Would you like to review the Puppeteer test?

@timabbott, @showell, please review.

akshatdalton commented on frontend_tests/puppeteer_tests/unsent-messages.ts:
```diff
@@ -0,0 +1,134 @@
+        }),
+    );
+
+    const zid = await page.evaluate(() => {
+        const row = zulip_test.last_visible_row();
+        return row.attr("zid");
+    });
+
+    const pattern = new RegExp(`\\d+\\.${id_ends_with}`);
+    assert(pattern.test(zid));
```
I remember `endsWith` was not working here, so I've to use RegExp this way.
